The goal of the tool is to make the process of identifying QTLs user friendly and easily repeatable.
The process is separated into two essential steps. The first step is to obtain **variant call format**
(`.vcf`) files. This is done by running the **Germline short variant discovery** pipeline from GATK.
The second step is to filter the data and to identify the _quantitative trait loci_ (QTL).
Both steps are simplified with a user interface created in **R** using **shiny**. For further
convince all tools and shiny server have been packed in to a **docker** container thus
eliminating the need for manual installation of tools included in the process.  

The data generated in the container is directly linked to the local files. The final and all the intermediate data generated by the pipeline tools can be found locally and will not be deleted even after closing the app or docker container. Locally edited data will be transferred onto the container. The app has the following folder structure:


    QTLspyer                            # Root project folder                   
    ├── shiny                           # Code for the R shiny application
    ├── variant_calling                 # Python scripts for variant calling pipeline
    ├── log                             # Process status reports and standard outputs from tools
    ├── input
    │   ├── adapters                    # Expects adapters for sequence trimming (.fasta)
    │   ├── annotation                  # Expects genome annotation (.gtf)
    │   ├── references                  # Expects genome references (.fasta & .vcf)
    │   └── sample_data                 # Expects sample read sequences (.fastq)
    └── output
        ├── aligned                     # Sequence alignment files location (.bam)
        ├── fastqc                      # FastQC Quality reports location (.html & .zip)
        ├── trimmed                     # Trimmed read sequences location (.fasqc)
        └── GATK                        # GATK tools for variant calling output location
            ├── VCFs                    # Germline Variant called files and their indexes (.vcf & .tbi)
            ├── nonfiltered             # SNPs selected out of variants and their indexes (.vcf & .idx)
            └── tables                  # VCF files transformed to tables (.snps.table)


QTL estimation is performed with R and uses the `.table` files found in `QTLspyer/output/GATK/tables/`. Those files can be removed or placed inside the folder as pleased.  

The variant calling pipeline and QTL estimations are standalone processes and can be performed separately without the need of repetition, being that the output files already exist.
